pkgs/jni/lib/src/jclass.dart

```diff
@@ -0,0 +1,154 @@
+extension type JInstanceField._fromPointer(JFieldIDPtr pointer) {
+  JInstanceField._(JClass jClass, String name, String signature)
+      : pointer = using((arena) => Jni.accessors
+            .getFieldID(
+              jClass.reference.pointer,
+              name.toNativeChars(arena),
+              signature.toNativeChars(arena),
+            )
+            .fieldID);
+
+  DartT get<JavaT, DartT>(JObject object, JAccessible<JavaT, DartT> type) {
+    return type._instanceGet(object.reference.pointer, this as JFieldIDPtr);
+  }
+
+  void set<JavaT, DartT>(
+      JObject object, JAccessible<JavaT, DartT> type, DartT value) {
+    type._instanceSet(object.reference.pointer, this as JFieldIDPtr, value);
+  }
+}
+
+/// A thin wrapper over a [JFieldIDPtr] of an static field.
+extension type JStaticField._fromPointer(JFieldIDPtr pointer) {
+  JStaticField._(JClass jClass, String name, String signature)
+      : pointer = using((arena) => Jni.accessors
+            .getStaticFieldID(
+              jClass.reference.pointer,
+              name.toNativeChars(arena),
+              signature.toNativeChars(arena),
+            )
+            .fieldID);
+
+  DartT get<JavaT, DartT>(JClass jClass, JAccessible<JavaT, DartT> type) {
+    return type._staticGet(jClass.reference.pointer, this as JFieldIDPtr);
+  }
+
+  void set<JavaT, DartT>(
+      JObject object, JAccessible<JavaT, DartT> type, DartT value) {
+    type._staticSet(object.reference.pointer, this as JFieldIDPtr, value);
+  }
+}
+
+/// A thin wrapper over a [JMethodIDPtr] of an instance method.
+extension type JInstanceMethod._fromPointer(JMethodIDPtr pointer) {
+  JInstanceMethod._(
+    JClass jClass,
+    String name,
+    String signature,
+  ) : pointer = using((arena) => Jni.accessors
+            .getMethodID(
+              jClass.reference.pointer,
+              name.toNativeChars(arena),
+              signature.toNativeChars(arena),
+            )
+            .methodID);
+
+  /// Calls the instance method on [object] with the given arguments.
+  DartT call<JavaT, DartT>(
+    JObject object,
+    JCallable<JavaT, DartT> returnType,
+    List<dynamic> args,
+  ) {
+    return using((arena) => returnType._instanceCall(object.reference.pointer,
+        this as JMethodIDPtr, toJValues(args, allocator: arena)));
+  }
+}
+
+/// A thin wrapper over a [JMethodIDPtr] of a static mehtod.
+extension type JStaticMethod._fromPointer(JMethodIDPtr pointer) {
+  JStaticMethod._(
+    JClass jClass,
+    String name,
+    String signature,
+  ) : pointer = using((arena) => Jni.accessors
+            .getStaticMethodID(
+              jClass.reference.pointer,
+              name.toNativeChars(arena),
+              signature.toNativeChars(arena),
+            )
+            .methodID);
+
+  /// Calls the static method on [jClass] with the given arguments.
+  DartT call<JavaT, DartT>(
+    JClass jClass,
+    JCallable<JavaT, DartT> returnType,
```

Comment:
Consider making this a class so that you can pass these at the constructor instead of every call site.

Or are these types used for calling generic methods?
In that case we would need to decide whether we'd like to have a new `JStaticMethod` for every invocation with different type args.
(The correct way to represent things would probably be to have the type argument bound be on the constructor and the invocation type argument to be on `call`.)

Reply:
`JStaticMethod` is just `JMethodIDPtr`, and that doesn't care about generics. So at least for `package:jnigen` bindings, this is the way. Hell, we might even consider that we actually don't want to have a final `JObject` and instead we want to get the `JReference` to save one type class creation. That's why we have `referenceType` – now only used for JNIgen constructors but could be used for other things as well.

https://github.com/dart-lang/native/pull/998/files#diff-9e1b96256e21a855546301263c6003e7a977def64572a57e68e123324c0220e2R46-R51